packages/safe-ds-vscode/src/extension/pythonServer.ts

```diff
@@ -0,0 +1,385 @@
+import child_process from 'child_process';
+import vscode from 'vscode';
+import net from 'net';
+import WebSocket from 'ws';
+import { ast, getModuleMembers, SafeDsServices } from '@safe-ds/lang';
+import { LangiumDocument, URI } from 'langium';
+import path from 'path';
+import { createProgramMessage, ProgramCodeMap, PythonServerMessage, RuntimeErrorBacktraceFrame } from './messages.js';
+import { logError, logOutput } from './output.js';
+import { BasicSourceMapConsumer, SourceMapConsumer } from 'source-map';
+
+let pythonServer: child_process.ChildProcessWithoutNullStreams | undefined = undefined;
+let pythonServerPort: number | undefined = undefined;
+let pythonServerAcceptsConnections: boolean = false;
+let pythonServerConnection: WebSocket | undefined = undefined;
+let pythonServerMessageCallbacks: Map<PythonServerMessage['type'], ((message: PythonServerMessage) => void)[]> =
+    new Map<PythonServerMessage['type'], ((message: PythonServerMessage) => void)[]>();
+
+/**
+ * Start the python server on the next usable port, starting at 5000.
+ * Uses the 'safe-ds.runner.command' setting to execute the process.
+ */
+export const startPythonServer = async function (): Promise<void> {
+    pythonServerAcceptsConnections = false;
+    const runnerCommandSetting = vscode.workspace.getConfiguration('safe-ds.runner').get<string>('command')!; // Default is set
+    const runnerCommandParts = runnerCommandSetting.split(/\s/u);
+    const runnerCommand = runnerCommandParts.shift()!; // After shift, only the actual args are left
+    // Test if the python server can actually be started
+    try {
+        const pythonServerTest = child_process.spawn(runnerCommand, [...runnerCommandParts, '-V']);
+        const versionString = await getPythonServerVersion(pythonServerTest);
+        logOutput(`Using safe-ds-runner version: ${versionString}`);
+    } catch (error) {
+        logError(`Could not start runner: ${error}`);
+        vscode.window.showErrorMessage('The runner process could not be started.');
+        return;
+    }
+    // Start the runner at the specified port
+    pythonServerPort = await findFirstFreePort(5000);
+    logOutput(`Trying to use port ${pythonServerPort} to start python server...`);
+    logOutput(`Using command '${runnerCommandSetting}' to start python server...`);
+
+    const runnerArgs = [...runnerCommandParts, '--port', String(pythonServerPort)];
+    logOutput(`Running ${runnerCommand}; Args: ${runnerArgs.join(' ')}`);
+    pythonServer = child_process.spawn(runnerCommand, runnerArgs);
+    manageSubprocessOutputIO(pythonServer);
+    try {
+        await connectToWebSocket();
+    } catch (error) {
+        stopPythonServer();
+    }
+    logOutput('Started python server successfully');
+};
+
+export const stopPythonServer = function (): void {
+    pythonServer?.kill();
+    // TODO maybe tree-kill sync process? Does not seem to always be reaped, and may remain as a zombie
```

Comment:
I thought about including the tree-kill dependency (https://www.npmjs.com/package/tree-kill) to clean up all started processes when stopping the extension.
tree-kill kills a process tree by executing commands in the shell, but should be reasonably cross-platform.

Reply:
Could you create an issue in the runner repo for this? Here, you can go ahead with the `tree-kill` solution for now.

Reply:
Since graceful shutdown signals are not supported on Windows, a new message type would be needed (like you proposed). 
(On Unix-like systems, SIGINT could be sent, which would interrupt the WSGI server)
So a cleanup operation that is always run (like registered with `atexit`) would not work (at least on Windows)